Fix(proto_change_detector): ensure that a dotted property is parsed correctly

matteosuppo opened the pull request:
Previously there was a conflict when the name of an object property (address.city)
was the same as the name of a local variable (city), and it evaluated to the local
variable. Now it detects if a property has a dot and returns the local variable
only if the dot is missing.

No breaking changes, fix issue #1542

matteosuppo committed:
Fix(proto_change_detector): ensure that a dotted property is parsed correctly

Previously there was a conflict when the name of an object property (address.city)
was the same as the name of a local variable (city), and it evaluated to the local
variable. Now it detects if a property has a dot and returns the local variable
only if the dot is missing.

No breaking changes, fix issue #1542

matteosuppo committed:
Chore(proto_change_detector): Compatibility with Dart

Travis wasn't happy with a non-boolean negation. I fixed it to check if
the variable is null instead.

caitp commented:
looks okay to me, maybe the presence of the '.' should be flagged in BindingRecord

vicb commented on modules/angular2/test/change_detection/change_detection_spec.js:
```diff
@@ -469,6 +469,18 @@ export function main() {
+              var address = new Address('Grenoble');
+              var person = new Person('Victor', address);
+              var locals = new Locals(null,
+                  MapWrapper.createFromPairs([["city", "Paris"]]));
```
nit pick: dont mix single & double quotes - & I would prefer MTV over Paris :)

matteosuppo replied:
I'll fix the quotes asap.

What's MTV? ò_O

vicb replied:
"Mountain View" but no need to worry about that ;)

matteosuppo committed:
style(test/change_detection): Use single quotes instead of double quotes

Also change Paris to MTV because why not?

vsavkin commented:
Maybe I am missing something, but as far as I can tell the PR does not fixe the issue.

I think you should fix it as follows:

```
 visitAccessMember(ast:AccessMember) {
    var receiver = ast.receiver.visit(this);
    if (isPresent(this.variableBindings) && ListWrapper.contains(this.variableBindings, ast.name) && ast.receiver instanceof ImplicitReceiver) { 
      return this._addRecord(RECORD_TYPE_LOCAL, ast.name, ast.name, [], null, receiver);
    } else {
      return this._addRecord(RECORD_TYPE_PROPERTY, ast.name, ast.getter, [], null, receiver);
    }
  }
```

matteosuppo committed:
style(proto_change_detector): Use instanceof instead of checking for dots

Checking for dots in the property name is ugly, while checking if the
receiver is a ImplicitReceiver instead of an AccessMember is better

mhevery commented:
d4925b6
